### ISSUE TITLE
Need Help getting AITool to invoke Custom

### ISSUE THREAD

**Issue by tommyjlong** [VorlonCD/bi-aidetection#251]
AITool Version 2.0.760
Deepstack Windows version

I have AITool controlling the startup of Deepstack Windows version.
It invokes object detection and face detection and it works fine.
I have recently added a Custom Model to Deepstack and have been trying to get AITool to use it.

I have added to AI AITool Deepstack Tab:
```
  Custom model (Checked)
  Paths(s): C:/Users/tommy/Documents/bin/DeepStack/MyModels
  Names(s): mailman
  Port(s): 84
```
In the Command line box it shows:
```
C:\DeepStack\server\DeepStack.exe --MODELSTORE-DETECTION "C:/Users/tommy/Documents/bin/DeepStack/MyModels" --PORT 84
C:\DeepStack\server\DeepStack.exe --VISION-FACE True --VISION-DETECTION True --MODE High --PORT 83
```
In the URL box it shows:
```
http://127.0.0.1:84/v1/vision/custom/mailman
http://127.0.0.1:83/v1/vision/detection
http://127.0.0.1:83/v1/vision/face/recognize
```

In the Settings Tab-> AI Server URLs
It shows these same 3 URLs in the order:
```
http://127.0.0.1:84/v1/vision/custom/mailman
http://127.0.0.1:83/v1/vision/detection
http://127.0.0.1:83/v1/vision/face/recognize
```

When AITool detects a new image and runs, looking at the logs it triggers calls to
```
http://127.0.0.1:83/v1/vision/detection
http://127.0.0.1:83/v1/vision/face/recognize
```
But it never triggers a call to
```
http://127.0.0.1:84/v1/vision/custom/mailman
```
I have restarted AITool, but the outcome is the same.

Any suggestions?

**Comment by VorlonCD** [comment]
pretty old version, try the latest:
https://github.com/VorlonCD/bi-aidetection/tree/master/src/UI/Installer

`/vision/detection` URL should be FIRST and you should pick the `LINK COMBINE RESULTS` checkbox, and check your MAILMAN one.   When you edit the mailman URL you might only want "`Use only as linked server`" checkbox. and nothing else.

Also I think the name of the custom model is case sensitive.

`face/recognize` URL should only have "`Use as refinement server`" checked and you should put "`person, people, mailman`" in the objects list below that so that this URL is not used when a car, dog or wookiee walks by.

**Comment by tommyjlong** [comment]
Thanks for this 👍 and thanks for your work on AITool!!

I believe I have the setup working correctly (well almost).  I did a few test runs and I see that AITool is now calling both the vision/detection and vision/custom/mailman and getting results back. 

This is what I have in the Edit AI Servers Window (in Order 1,2,3 respectively):
1. DeepStack  "Link/Combine Results with other servers";  Wait for results from all link servers...:
              http://127.0.0.1:84/v1/vision/custom/mailmain
2. DeepStack_Custom "Use ONLY as linked server"
3. DeepStack_Faces "Use a Refinement Server"
                   Use this server ONLY if another server detects the following objects first:
                     person,face.

However, today when the mailman actually came by (mailman in this case is actually a vehicle that my custom renders as `mailtruck`),  it appears that /vision/detection labeled the vehicle as `Bus`, and the vision/custom/mailman render the same object correctly as `mailtruck`, but when AITool combined them, it tossed the Mailtruck.

When I look at the History->"Prediction Object Detail" it has two entries:
1) Bus entry is green. 
2) Mailtruck entry is gray'd out (which I understand to mean the entry is not used), and I'm not sure why.

From the corresponding debug log it says:
```
Result='Relevant', Detail='Bus (79%)', ObjType='Vehicle', ObjectResult=Relevant, DynMaskResult='NotEnabled', DynMaskType='None', ImgMaskResult='MajorityOutsideMask', ImgMaskType='Image, PercentOfImage=0.6%|5184|1|Red|30
Result='RelevantDuplicateObject', Detail='Mailtruck (49%)', ObjType='Vehicle', ObjectResult=Relevant, DynMaskResult='SkippedDynamicMaskCheck', DynMaskType='None', ImgMaskResult='MajorityOutsideMask', ImgMaskType='Image'|5185|1|Silver|30
```
When the MQTT message was sent out, it contained the `Bus` but not the `Mailtruck`. For this camera, I have set the MQTT Mailtruck check box.

Here is the entire debug log for this transaction:
```
2021-06-08 11:07:55.799183|Debug|AITOOLS.EXE|GetCamera|Deepstack_Objects:127.0.0.1:83|FrontYard||   *** Note: More than one configured camera matched '\\192.168.0.33\public\Front20210608_110752M.jpg', using the first one matched: 'FrontYard' (StartsWith-Prefix) ***|5138|2||26
2021-06-08 11:07:55.799881|Debug|AITOOLS.EXE|AddImageToQueue|Deepstack_Objects:127.0.0.1:83|FrontYard||  |5143|1||26
2021-06-08 11:07:55.799881|Debug|AITOOLS.EXE|AddImageToQueue|Deepstack_Objects:127.0.0.1:83|FrontYard|Front20210608_110752M.jpg|  ====================== Adding new image to queue (Count=1): \\192.168.0.33\public\Front20210608_110752M.jpg|5144|1||26
2021-06-08 11:07:55.803930|Info|AITOOLS.EXE|AddImageToQueue|Deepstack_Objects:127.0.0.1:83|FrontYard||  Skipping image because of duplicate Created File Event: \\192.168.0.33\public\Front20210608_110752M.jpg|5145|2||26
2021-06-08 11:07:55.838113|Debug|AITOOLS.EXE|GetCamera|Deepstack_Objects:127.0.0.1:83|FrontYard||  *** Note: More than one configured camera matched '\\192.168.0.33\public\Front20210608_110752M.jpg', using the first one matched: 'FrontYard' (StartsWith-Prefix) ***|5146|1||11
2021-06-08 11:07:55.839112|Debug|AITOOLS.EXE|WaitForNextURL|Deepstack_Objects:127.0.0.1:83|FrontYard||  ---- Found '1' linked AI URL's.|5151|1||11
2021-06-08 11:07:55.839112|Debug|AITOOLS.EXE|ImageQueueLoop|Deepstack_Objects:127.0.0.1:83|FrontYard|Front20210608_110752M.jpg| Adding task for file 'Front20210608_110752M.jpg' (Image QueueTime='0.0' mins, URL Queue wait='0ms', URLOrder=2, URLOriginalOrder=1) on URL 'http://127.0.0.1:83/v1/vision/detection'|5152|0||11
2021-06-08 11:07:55.840113|Debug|AITOOLS.EXE|ImageQueueLoop|Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg| Adding task for file 'Front20210608_110752M.jpg' (Image QueueTime='0.0' mins, URL Queue wait='0ms', URLOrder=2, URLOriginalOrder=2) on URL 'http://127.0.0.1:84/v1/vision/custom/mailman'|5153|0||11
2021-06-08 11:07:55.840113|Debug|AITOOLS.EXE|ImageQueueLoop|Deepstack_Custom:127.0.0.1:84|FrontYard|| Done adding 2 total threads, ErrCnt=0, ImageProcessQueue.Count=0|5154|0||11
2021-06-08 11:07:55.840113|Debug|AITOOLS.EXE|GetCamera|Deepstack_Custom:127.0.0.1:84|FrontYard||   *** Note: More than one configured camera matched '\\192.168.0.33\public\Front20210608_110752M.jpg', using the first one matched: 'FrontYard' (StartsWith-Prefix) ***|5155|2||59
2021-06-08 11:07:55.840113|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|   Starting analysis of \\192.168.0.33\public\Front20210608_110752M.jpg...|5160|2||59
2021-06-08 11:07:56.289524|Debug|AITOOLS.EXE|LoadImage|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard||   Image file is valid. Resolution=1920x1536, LockMS=385ms (max=10000ms), retries=6, size=407 KB: Front20210608_110752M.jpg|5161|2||59
2021-06-08 11:07:56.290516|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|  (Image resolution=1920x1536 @ 96 DPI and 407 KB)|5162|1||59
2021-06-08 11:07:56.292695|Debug|AITOOLS.EXE|GetDetectionsFromAIServer|Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|    (1/6) Uploading a 407 KB image to 'DeepStack_Custom' (CAM_LowerThresholdOverride=0,Upper=100) AI Server at http://127.0.0.1:84/v1/vision/custom/mailman|5163|3||59
2021-06-08 11:07:56.292695|Debug|AITOOLS.EXE|GetDetectionsFromAIServer|Deepstack_Objects:127.0.0.1:83|FrontYard|Front20210608_110752M.jpg|    (1/6) Uploading a 407 KB image to 'DeepStack' (CAM_LowerThresholdOverride=0,Upper=100) AI Server at http://127.0.0.1:83/v1/vision/detection|5164|3||11
2021-06-08 11:07:56.753432|Debug|AITOOLS.EXE|DSHandleServerProcMSG|Deepstack_Objects:127.0.0.1:83|FrontYard|DEEPSTACK.EXE|    DeepStack:84>> [GIN] 2021/06/08 - 11:07:56 ; 200 ;    460.7389ms ;       127.0.0.1 ; POST     /v1/vision/custom/mailman|5165|3||25
2021-06-08 11:07:57.184184|Debug|AITOOLS.EXE|DSHandleServerProcMSG|Deepstack_Objects:127.0.0.1:83|FrontYard|DEEPSTACK.EXE|   DeepStack:83>> [GIN] 2021/06/08 - 11:07:57 ; 200 ;     891.492ms ;       127.0.0.1 ; POST     /v1/vision/detection|5166|2||41
2021-06-08 11:07:57.184184|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Objects:127.0.0.1:83|FrontYard|Front20210608_110752M.jpg|  (2/6) [Primary] Posted in 891ms, StatusCode='OK', Received a 135 byte JSON response: '{"success":true,"predictions":[{"confidence":0.78516346,"label":"bus","y_min":144,"x_min":1108,"y_max":250,"x_max":1288}],"durat...'|5167|1||30
2021-06-08 11:07:57.184184|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Objects:127.0.0.1:83|FrontYard|Front20210608_110752M.jpg|  (3/6) [Primary] Processing 1 results...|5168|1||30
2021-06-08 11:07:57.184184|Debug|AITOOLS.EXE|IsRelevant|Deepstack_Objects:127.0.0.1:83|FrontYard||    RelevantObjectManager: Object is valid for 'Default (NEW)' because:  Relevant='Bus',. , 1 predictions(s), Enabled=13 of 14, IsNew=True|5169|3||30
2021-06-08 11:07:57.193179|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Objects:127.0.0.1:83|FrontYard|Front20210608_110752M.jpg|  (2/6) [Linked] Posted in 461ms, StatusCode='OK', Received a 141 byte JSON response: '{"success":true,"predictions":[{"confidence":0.49136272,"label":"mailtruck","y_min":137,"x_min":1100,"y_max":252,"x_max":1296}],...'|5172|1||30
2021-06-08 11:07:57.193179|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Objects:127.0.0.1:83|FrontYard|Front20210608_110752M.jpg|  (3/6) [Linked] Processing 1 results...|5173|1||30
2021-06-08 11:07:57.193179|Debug|AITOOLS.EXE|IsRelevant|Deepstack_Objects:127.0.0.1:83|FrontYard||    RelevantObjectManager: Object is valid for 'Default (NEW)' because:  Relevant='Mailtruck',. , 1 predictions(s), Enabled=13 of 14, IsNew=True|5174|3||30
2021-06-08 11:07:57.202175|Debug|AITOOLS.EXE|WaitForNextURL|Deepstack_Objects:127.0.0.1:83|FrontYard||   ---- Refinement server requested, but none were available for predictions 'Bus,Mailtruck,'. (0ms)|5177|2||30
2021-06-08 11:07:57.202175|Debug|AITOOLS.EXE|IsRelevant|Deepstack_Objects:127.0.0.1:83|FrontYard||    RelevantObjectManager: Object is valid for 'Default (NEW)' because:  Relevant='Bus',. , 1 predictions(s), Enabled=13 of 14, IsNew=True|5180|3||30
2021-06-08 11:07:57.210169|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|     Detected objects:|5183|1||30
2021-06-08 11:07:57.210169|Info|AITOOLS.EXE|DetectObjects|Deepstack_Objects:127.0.0.1:83|FrontYard|Front20210608_110752M.jpg|      Result='Relevant', Detail='Bus (79%)', ObjType='Vehicle', ObjectResult=Relevant, DynMaskResult='NotEnabled', DynMaskType='None', ImgMaskResult='MajorityOutsideMask', ImgMaskType='Image, PercentOfImage=0.6%|5184|1|Red|30
2021-06-08 11:07:57.210169|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|      Result='RelevantDuplicateObject', Detail='Mailtruck (49%)', ObjType='Vehicle', ObjectResult=Relevant, DynMaskResult='SkippedDynamicMaskCheck', DynMaskType='None', ImgMaskResult='MajorityOutsideMask', ImgMaskType='Image'|5185|1|Silver|30
2021-06-08 11:07:57.210169|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|  The summary:Bus (79%)|5186|1||30
2021-06-08 11:07:57.210169|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|  (5/6) Performing alert actions:|5187|1||30
2021-06-08 11:07:57.210169|Debug|AITOOLS.EXE|MergeImageAnnotations|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|      Merging image annotations: \\192.168.0.33\public\Front20210608_110752M.jpg|5188|5||30
2021-06-08 11:07:57.243150|Debug|AITOOLS.EXE|MergeImageAnnotations|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|      Merged 1 detections in 32ms into image C:\Users\tommy\AppData\Local\Temp\Front20210608_110752M.jpg|5189|5||30
2021-06-08 11:07:57.270136|Debug|AITOOLS.EXE|LoadImage|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard||      Image file is valid. Resolution=1920x1536, LockMS=5ms (max=10000ms), retries=0, size=441 KB: Front20210608_110752M.jpg|5190|5||30
2021-06-08 11:07:57.271135|Debug|AITOOLS.EXE|IsRelevant|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard||      RelevantObjectManager: Object is valid for 'MQTT' because:  Relevant='Bus',Irrelevant='Mailtruck',. , 2 predictions(s), Enabled=21 of 22, IsNew=False|5191|5||30
2021-06-08 11:07:57.272135|Debug|AITOOLS.EXE|Trigger|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard||     MQTT Trigger event - [SummaryNonEscaped]='Bus (79%)', After replacement Topic='ai/FrontYard/motion', Payload='{"fileName":"\\\\192.168.0.33\\public\\front20210608_110752m.jpg","baseName":"front20210608_110752m.jpg","summary":"Bus (79%)","state":"on","analysisDurationMS":1352,"time":"2021-06-08T11:07:57.2101695-04:00","camera":"FrontYard","predictions":[{"label":"Bus","detail":"","userID":"","confidence":78.516346,"y_min":144.0,"x_min":1108.0,"y_max":250.0,"x_max":1288.0,"server":"Deepstack_Objects:127.0.0.1:83"}]}'|5192|4||30
2021-06-08 11:07:57.272135|Debug|AITOOLS.EXE|PublishAsync|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard||      MQTT: Sending topic 'ai/FrontYard/motion' with payload '{"fileName":"\\\\192.168.0.33\\public\\front20210608_110752m.jpg","baseName":"front20210608_110752m.jpg","summary":"Bus (79%)","state":"on","analysisDurationMS":1352,"time":"2021-06-08T11:07:57.2101695-04:00","camera":"FrontYard","predictions":[{"label":"Bus","detail":"","userID":"","confidence":78.516346,"y_min":144.0,"x_min":1108.0,"y_max":250.0,"x_max":1288.0,"server":"Deepstack_Objects:127.0.0.1:83"}]}' to server '192.168.0.34:1883'...|5193|5||41
2021-06-08 11:07:57.273134|Debug|AITOOLS.EXE|PublishAsync|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard||      MQTT: ...Sent in 0ms, Reason: 'Success' (0 - '')|5194|5||11
2021-06-08 11:07:57.273134|Debug|AITOOLS.EXE|Trigger|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|     FrontYard last triggered at 6/8/2021 11:07:57 AM.|5195|4||11
2021-06-08 11:07:57.273134|Debug|AITOOLS.EXE|RunTriggers|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|     Cancel action 'All' queued due to event 'Bus (79%)' for camera 'FrontYard', waiting 30 seconds...|5196|4||11
2021-06-08 11:07:57.273134|Debug|AITOOLS.EXE|RunTriggers|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|     Action 'Cancel' done. Succeeded=True, Trigger=False, Queued=False, Queue Count=0 (Min=0,Max=0,Avg=0), Total time=63ms (Min=2ms,Max=32800ms,Avg=2932ms), Queue time=0 (Min=6ms,Max=32797ms,Avg=26618ms), Action Time=62ms (Min=1ms,Max=212ms,Avg=20ms), Image=\\192.168.0.33\public\Front20210608_110752M.jpg|5197|4||11
2021-06-08 11:07:57.273134|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|     (6/6) SUCCESS.|5198|4||11
2021-06-08 11:07:57.273134|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|              Total Time:   1473 ms (Count=    58, Min=   845 ms, Max=  4067 ms, Avg=  1852 ms)|5199|4||11
2021-06-08 11:07:57.273134|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Objects:127.0.0.1:83|FrontYard|Front20210608_110752M.jpg|           AI (URL) Time:    891 ms (Count=    58, Min=     1 ms, Max= 55047 ms, Avg=   739 ms)|5200|4||11
2021-06-08 11:07:57.273134|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|           AI (URL) Time:    461 ms (Count=    58, Min=   336 ms, Max=  1072 ms, Avg=   435 ms)|5201|4||11
2021-06-08 11:07:57.273134|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|          File lock Time:    385 ms (Count=    58, Min=     9 ms, Max=   453 ms, Avg=    94 ms)|5202|4||11
2021-06-08 11:07:57.273134|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|          File load Time:     61 ms (Count=    58, Min=     3 ms, Max=   225 ms, Avg=    21 ms)|5203|4||11
2021-06-08 11:07:57.273134|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|        Image Queue Time:     40 ms (Count=    58, Min=     3 ms, Max=  2723 ms, Avg=   811 ms)|5204|4||11
2021-06-08 11:07:57.273134|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|         Image Life Time:   1710 ms (Count=    58, Min=  1062 ms, Max=  4352 ms, Avg=  2106 ms)|5205|4||11
2021-06-08 11:07:57.273134|Debug|AITOOLS.EXE|DetectObjects|Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|       Image Queue Depth:      1    (Count=    58, Min=     1,    Max=     1,    Avg=     1)|5206|4||11
2021-06-08 11:07:57.288123|Debug|AITOOLS.EXE|ScanImages|Deepstack_Custom:127.0.0.1:84|FrontYard||    Found 0 Default*.jpg files in \\192.168.0.33\public|5207|3||11
2021-06-08 11:07:57.289123|Debug|AITOOLS.EXE|ScanImages|Deepstack_Custom:127.0.0.1:84|FrontYard||    Found 0 images in 11 ms. Scanning images...|5208|3||11
2021-06-08 11:07:57.289123|Debug|AITOOLS.EXE|ScanImages|Deepstack_Custom:127.0.0.1:84|FrontYard||    0 of 0 image files processed, 0 new resolutions found (0 invalid) in 0 ms (Max=500 ms), 0 different image resolutions found: |5209|3||11
2021-06-08 11:07:57.306114|Debug|AITOOLS.EXE|SaveAsync|Deepstack_Custom:127.0.0.1:84|FrontYard||  JSON Settings saved to REGISTRY and C:\AITOOL\_Settings\AITool.Settings.JSON|5210|1||41
2021-06-08 11:07:59.488200|Debug|AITOOLS.EXE|LoadImage|Deepstack_Custom:127.0.0.1:84|FrontYard||  Image file is valid. Resolution=1920x1536, LockMS=0ms (max=10000ms), retries=0, size=407 KB: front20210608_110752m.jpg|5211|1||1
2021-06-08 11:08:27.774269|Debug|AITOOLS.EXE|CancelActionJobQueueLoop|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg| Running cancel Action 'Cancel' in queue for event 'Bus (79%)' for camera 'FrontYard', after 30.5 seconds...|5212|0||17
2021-06-08 11:08:27.777296|Debug|AITOOLS.EXE|Trigger|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard||    MQTT Cancel event - [SummaryNonEscaped]='Bus (79%)', After replacement Topic='ai/FrontYard/motion', Payload='cancel'|5213|3||17
2021-06-08 11:08:27.777296|Debug|AITOOLS.EXE|PublishAsync|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard||     MQTT: Sending topic 'ai/FrontYard/motion' with payload 'cancel' to server '192.168.0.34:1883'...|5214|4||17
2021-06-08 11:08:27.778273|Debug|AITOOLS.EXE|PublishAsync|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard||     MQTT: ...Sent in 0ms, Reason: 'Success' (0 - '')|5215|4||8
2021-06-08 11:08:27.778273|Debug|AITOOLS.EXE|RunTriggers|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|  Cancel action 'Cancel' has been removed due to event 'Bus (79%)' for camera 'FrontYard'|5216|1||8
2021-06-08 11:08:27.778273|Debug|AITOOLS.EXE|RunTriggers|Deepstack_Objects:127.0.0.1:83;Deepstack_Custom:127.0.0.1:84|FrontYard|Front20210608_110752M.jpg|  Action 'Cancel' done. Succeeded=True, Trigger=False, Queued=False, Queue Count=0 (Min=0,Max=0,Avg=0), Total time=30568ms (Min=2ms,Max=32800ms,Avg=3357ms), Queue time=30564 (Min=6ms,Max=32797ms,Avg=27112ms), Action Time=3ms (Min=1ms,Max=212ms,Avg=20ms), Image=\\192.168.0.33\public\Front20210608_110752M.jpg|5217|1||8
```

**Comment by VorlonCD** [comment]
From a quick look it seems to be working as designed.   it took two overlapping objects and merged into one - That one being the BUS because it has a higher confidence than the mail truck.   If the bus had a higher confidence I believe it would have won.

I suspect the only fix is to uncheck BUS so mailtruck is the only thing found.

But you could first try moving mailtruck above bus in the relevant objects list, and maybe move "Deepstack_custom" above "Deepstack".  But I seem to remember it doesnt help with sorting/merging dupes much.

When a dupe does get merged/ignored, this version will now make sure the DETAILS property/column has the name of the merged objects if different.  So in your JSON sent via mqtt you should be able to see "mailtruck"
https://github.com/VorlonCD/bi-aidetection/tree/master/src/UI/Installer

**Comment by iankaufmann** [comment]
@tommyjlong is your `mailman` model something you made yourself?  Very interested...

**Comment by tommyjlong** [comment]
@VorlonCD 
I.  Relevant Objects List
I tried out your suggestion moving "mailtruck" to the top of the "Relevant Objects" list.  But no change.

II. "Truck" versus "Mailtruck"
Another interesting case:  Yesterday, Deepstack vision/detection detected the mailman vehicle as a "Truck" (instead of a "Bus"), and the vision/custom/mailman detected the same object as "Mailtruck" as expected.  However the bounding box of the "Truck" versus the bounding box of the "Mailtruck" was different enough that when AITool tried to merge the two, there was NO duplicate declared.  So things worked wonderfully as it declared both "Mailtruck" and "Truck" and my MQTT message got both as separate "labels" !!

III. 2.0.1158 Version
I tried out the 2.0.1158 version.  For the duplicated case ("Bus" and "Mailtruck" declared as duplicate but "Bus" had higher confidence than "Mailtruck" so "Mailtruck" was tossed out), the new MQTT message (I'm using AllJson) contains the new detail:
`...[{"label":"Bus","detail":"Mailtruck","userID":"","confidence":78.516346,..."}`

I can parse this and make it work.  However, the `Mailtruck` confidence percentage is lost and I would like to keep using it.  

For MQTT, would it be possible to include both the "Bus" and the "Mailtruck" as separate labels, thus keeping their confidence levels, and if needed, flag the "Mailtruck" as Duplicate.  The MQTT message might look something like:
`....[{"label":"Bus","detail":"Mailtruck","userID":"","confidence":74.65194000000001,......"},  {"label":"Mailtruck","detail":"RelevantDuplicateObject","userID":"","confidence":68.10548,....."}]....`

Thanks!

**Comment by tommyjlong** [comment]
@iankaufmann 
Yes, when the mailman vehicle comes by to drop off the mail, my surveillance camera's motion detection captures an image of this.  The motion detection is not always consistent as to when it triggers (sometimes early, sometimes late), so you get somewhat differing angles of the vehicle and differing levels of brightness (sunny) or dull (cloudy) over several days.  I kept oh around two weeks worth of these images to use with Deepstack's custom model.  After that it was pretty much follow the instructions in the Deepstack documentation.  I used the Google CoLab to train, as I don't have a GPU.  Here is a [DeepStack YouTube](https://www.youtube.com/watch?v=wQKUQ6Y2n3Q) if interested.

**Comment by iankaufmann** [comment]
> @iankaufmann
> Yes, when the mailman vehicle comes by to drop off the mail, my surveillance camera's motion detection captures an image of this. The motion detection is not always consistent as to when it triggers (sometimes early, sometimes late), so you get somewhat differing angles of the vehicle and differing levels of brightness (sunny) or dull (cloudy) over several days. I kept oh around two weeks worth of these images to use with Deepstack's custom model. After that it was pretty much follow the instructions in the Deepstack documentation. I used the Google CoLab to train, as I don't have a GPU. Here is a [DeepStack YouTube](https://www.youtube.com/watch?v=wQKUQ6Y2n3Q) if interested.

Ah, cool.  I started trying to do that one night on Google CoLab but I definitely didn't have enough images.  I was hoping you had found a more generic library that did mail truck identification.

I'll have to give it another try!

I have started working on something using the Sighthound ALPR docker container that they offer for free.  It does happen to work really well for USPS, UPS, and FedEx (and most vehicle make/mode/color).  It just doesn't have any way to integrate with APIs/webhooks or any other method of getting the data back out.  

However, I'm able to see the info in some log files and am working on some scripts that will exfiltrate that data back out and get it into Blue Iris / Home Assistant.

That said... it would be ideal if this was something that worked with Deepstack / AI Tool without the extra hoops!

**Comment by VorlonCD** [comment]
Ok, @tommyjlong - the new version makes a few changes that should help.

- If 2 relevant objects are overlapping, the **relevant object order** now has priority over Confidence levels when removing duplicates.  So first just try moving mailtruck above truck/van/etc.
- Ability to disabling merging duplicate objects in **AITOOL.SETTINGS.JSON** via '`HistoryMergeDuplicatePredictions`'=False.  Defaults to true.

https://github.com/VorlonCD/bi-aidetection/tree/master/src/UI/Installer

**Comment by tommyjlong** [comment]
Thanks!!
I tried both options and they both work.  I decided to go with the `'HistoryMergeDuplicatePredictions=False`

Again Thanks for your effort on this project and helping me out on this :)
You may close this.

**Comment by kpally4** [comment]
thanks to this explanation, i'm able to run a detection+custom too :)
![hIWeTvd](https://user-images.githubusercontent.com/76535029/121949015-47c76c80-cd58-11eb-9440-f11fdd2b0f1f.png)

**Comment by tkffaul** [comment]
@VorlonCD  I also have an issue here... I have a couple objects that overlap... I trained the AI so see the objects and that works great, but every now and then they are perceived as a person (right amount of sunlight what have you). I've moved the objects up on the list (above person), but it shows all 3 objects and flags as person. Is there a way to stop it from combining the objects and just use the individual results?
